reactor-core/src/main/java/reactor/core/publisher/Operators.java

```diff
@@ -2574,6 +2574,205 @@ private LiftFunction(@Nullable Predicate<Publisher> filter,
 		}
 	}
 
+	/*package*/ static class MonoInnerProducerBase<O>
+			implements InnerProducer<O> {
+
+		private final CoreSubscriber<? super O> actual;
+
+		/**
+		 * The value stored by this Mono operator.
+		 */
+		@Nullable
+		private volatile O   value;
+
+		private volatile  int state; //see STATE field updater
+		@SuppressWarnings("rawtypes")
+		private static final AtomicIntegerFieldUpdater<MonoInnerProducerBase> STATE =
+				AtomicIntegerFieldUpdater.newUpdater(MonoInnerProducerBase.class, "state");
+
+		public MonoInnerProducerBase(CoreSubscriber<? super O> actual) {
+			this.actual = actual;
+		}
+
+		@Override
+		@Nullable
+		public Object scanUnsafe(Attr key) {
+			if (key == Attr.CANCELLED) return isCancelled();
+			if (key == Attr.TERMINATED) return (state & HAS_COMPLETED) == HAS_COMPLETED ;
+			if (key == Attr.PREFETCH) return Integer.MAX_VALUE;
+			return InnerProducer.super.scanUnsafe(key);
+		}
+
+		/**
+		 * Tries to emit the value and complete the underlying subscriber or
+		 * stores the value away until there is a request for it.
+		 * <p>
+		 * Make sure this method is called at most once
+		 * @param v the value to emit
+		 */
+		public final void complete(O v) {
+			doOnComplete(v);
+			while(true) {
+				int s = this.state;
+				if (s == CANCELLED) {
+					discard(v);
+					return;
+				}
+
+				if ((s & HAS_REQUEST) == HAS_REQUEST && STATE.compareAndSet(this, s, s | (HAS_VALUE | HAS_COMPLETED))) {
```

Comment:
let's create static methods such as `boolean hasRequest(long state)` and of others bit check so the code becomes much more readable (at least that what I usually do)